graphql-ballerina/duplicate_field_remover.bal

```diff
@@ -0,0 +1,194 @@
+        self.removeDuplicateSelections(fragmentNode.getSelections());
+        self.removeDuplicateFields(fragmentNode.getFields());
+        self.removeDuplicateFragments(fragmentNode.getFragments());
+        foreach parser:Selection selection in fragmentNode.getSelections() {
+            self.visitSelection(selection);
+        }
+    }
+
+    public isolated function visitArgument(parser:ArgumentNode argumentNode, anydata data = ()) {
+        // Do nothing
+    }
+
+    private isolated function removeDuplicateSelections(parser:Selection[] selections) {
+        map<parser:Selection> visitedSelections = {};
+        map<parser:Selection> visitedFragmentOnTypes = {};
+        parser:Selection[] duplicateSelections = [];
+        foreach parser:Selection selection in selections {
+            if(selection.isFragment) {
+                parser:FragmentNode fragmentNode = self.documentNode.getFragments().get(selection.name);
+                if (visitedFragmentOnTypes.hasKey(fragmentNode.getOnType())) {
+                    if (fragmentNode.isInlineFragment()) {
+                        duplicateSelections.push(selection);
+                    } else {
+                        self.appendDuplicateSelections(selection, visitedFragmentOnTypes.get(fragmentNode.getOnType()));
+                    }
+                } else {
+                    visitedFragmentOnTypes[fragmentNode.getOnType()] = selection;
+                }
+            } else {
+                if (visitedSelections.hasKey(selection.name)) {
+                    self.appendDuplicateSelections(selection, visitedSelections.get(selection.name));
+                    duplicateSelections.push(selection);
+                } else {
+                    visitedSelections[selection.name] = selection;
+                }
+            }
+        }
+        int i = 0;
+        while (i < selections.length()) {
+            foreach parser:Selection duplicate in  duplicateSelections {
+                if (selections[i] === duplicate) {
+                    parser:Selection remove = selections.remove(i);
+                    i -= 1;
+                }
+            }
+            i += 1;
+        }
+    }
+
+    private isolated function removeDuplicateFields(parser:FieldNode[] fields) {
+        string[] visitedFields = [];
+        parser:FieldNode[] duplicateFields = [];
+        foreach parser:FieldNode fieldNode in fields {
+            if (visitedFields.indexOf(fieldNode.getName()) == ()) {
+                visitedFields.push(fieldNode.getName());
+            } else {
+                duplicateFields.push(fieldNode);
+            }
+        }
+        int i = 0;
+        while (i < fields.length()) {
+            foreach parser:FieldNode duplicate in  duplicateFields {
+                if (fields[i] === duplicate) {
+                    parser:FieldNode remove = fields.remove(i);
+                    i -= 1;
+                }
+            }
+            i += 1;
+        }
+    }
```

Comment:
Can't we use a single loop here?

Maybe using a map to keep the duplicates? I feel like this logic can be improved to log(n)

Reply:
Can not have a unique key for each duplicates field node. So that it's hard to use a map here.

Reply:
Any suggestions?

Reply:
Can't we use `arrays:filter()` here?

Reply:
We need two loop if we use ```arrays:filter( )```, don't we? one for filter function and one for put filtered values into ```fields``` ?

Reply:
This is what I have in mind:

```ballerina
private isolated function removeDuplicateFields(parser:ParentNode node) {
    parser:fieldNode[] fields = node.getFields();
    string[] visitedFields = [];
    parser:FieldNode[] duplicateFields = [];
    foreach parser:FieldNode fieldNode in fields {
        if (visitedFields.indexOf(fieldNode.getName()) == ()) {
             visitedFields.push(fieldNode.getName());
        } else {
            duplicateFields.push(fieldNode);
        }
    }
    int i = 0;

    // Filter the fieldsArray here, and assign it back to the node.
    node.fields = filteredArray 
}
```

You may have to introduce a new function in the `parentNode` class to set the fields.